internal/controller/hubble/hubble.go

```diff
@@ -0,0 +1,154 @@
+	// Validate PEM data
+	if !isValidPEM(caCertData) || !isValidPEM(clientCertData) || !isValidPEM(clientKeyData) {
+		return nil, errors.New("invalid PEM data: ensure proper headers and formatting")
+	}
+
+	// Load client certificate and key from byte slices
+	cert, err := tls.X509KeyPair(clientCertData, clientKeyData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load client certificate and key from data: %w", err)
+	}
+	logger.Debug("Client certificate and key loaded from data.")
+
+	// Load CA certificate from byte slice
+	certPool := x509.NewCertPool()
+	if !certPool.AppendCertsFromPEM(caCertData) {
+		return nil, errors.New("failed to append CA certificate to pool from data")
+	}
+	logger.Debug("CA certificate appended to pool from data.")
+
+	tlsConfig := &tls.Config{
+		ServerName:         hubbleRelayExpectedServerName,
+		Certificates:       []tls.Certificate{cert},
+		RootCAs:            certPool,
+		InsecureSkipVerify: false,
+		MinVersion:         tls.VersionTLS12,
+	}
+
+	var cred credentials.TransportCredentials
+	if aksManaged {
+		logger.Debug("AKS Managed: Using experimental NewTLSWithALPNDisabled")
+		cred = exp_credentials.NewTLSWithALPNDisabled(tlsConfig, logger)
+	}
+
+	return cred, nil
+}
+
+// GetHubbleTransportCredentials read TLS transport credentials to connect to Hubble from the given Secret.
+func GetHubbleCTransportCredentials(ctx context.Context, clientset kubernetes.Interface,
+	logger *zap.Logger, secretName, namespace string, isAKSManaged bool) (credentials.TransportCredentials, error) {
+	// Fetch mTLS certificates
+	caData, clientCertData, clientKeyData, fetchErr := getHubbleMTLSCertificatesFromSecret(
+		ctx, clientset, logger, secretName, namespace,
+	)
+	if fetchErr != nil {
+		return nil, fmt.Errorf("failed to fetch mTLS certificates: %w", fetchErr)
+	}
+
+	// Load credentials
+	creds, loadErr := loadMTLSCredentialsFromData(logger, isAKSManaged, caData, clientCertData, clientKeyData)
+	if loadErr != nil {
+		return nil, fmt.Errorf("failed to load mTLS credentials: %w", loadErr)
+	}
+
+	return creds, nil
+}
+
+// isValidPEM validates if the given data is properly PEM-encoded.
+func isValidPEM(data []byte) bool {
+	block, _ := pem.Decode(data)
+	return block != nil
+}
+
+// DiscoverCiliumHubbleRelayAddress uses a kubernetes clientset in order to discover the address of the hubble-relay se…
+func DiscoverCiliumHubbleRelay(ctx context.Context, ciliumNamespace string, clientset kubernetes.Interface) (*v1.Service, error) {
+	service, err := clientset.CoreV1().Services(ciliumNamespace).Get(ctx, ciliumHubbleRelayServiceName, metav1.GetOptions{})
+	if err != nil {
+		return nil, ErrHubbleNotFound
+	}
+
+	return service, nil
+}
+
+// GetHubbleRelayAddress returns the address of the hubble-relay service to connect a gRPC client to.
+func GetHubbleRelayAddress(service *v1.Service) (string, error) {
```

Comment:
This doesn't seem to be specific to Hubble.
```suggestion
// GetAddressFromService returns the address of the given service to connect a gRPC client to.
func GetAddressFromService(service *v1.Service) (string, error) {
```